### PR TITLE
Add deque for JavaScript and typescript

### PR DESCRIPTION
**Pull request by zhuoqinyue**
If this PR is related to coding or code translation, please fill out the checklist.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

**Review by krahets — COMMENTED** [review]
Do js and ts doesn't provide built-in Deque? If so, please implement Deque using Linked List, because the `popFront()` is slow if implementing Deque using Array.

**Comment by zhuoqinyue** [comment]
OK. I would change the implementation

**Comment by krahets** [comment]
> OK. I would change the implementation

Or implement it using circular array (a little more complicated than using linked list).

**Comment by krahets** [comment]
Hi, we've added `linkedlist_deque.java` and you can refer to it.

I think you can directly use Array as the built-in deque in JS and TS, and add a comment to reveal O(n) time complexity for the `addFirst()` and `pollFirst()` methods.

**Comment by zhuoqinyue** [comment]
Fine. I've enjoyed my Spring Festival vacation recently. I just returned to the city where I live today.
And I would implement Deque using `DequeNode` as follow:
```
class DequeNode {
    prev = null; 
    next = null;
    val: number = 0;
}
```
If I has no idea to implement Deque, I will refer to `linkedlist_deque.java` .

**Comment by zhuoqinyue** [comment]
@krahets hey, I've commented the time complexity. 
And I will implement Deque using `DequeNode` for `linkedlist_deque`

File: codes/javascript/chapter_stack_and_queue/deque.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    len;
+    constructor() {
+        this.len = 0;
+        this.array = [];
+    }
+
+    /* 在队尾添加元素 */
+    pushBack(value) {
+        this.array.push(value);
+        this.len++;
```
Comment (krahets) [inline codes/javascript/chapter_stack_and_queue/deque.js:19]: I think it is not necessary to record `len`. We can get it from `array.length`

File: codes/javascript/chapter_stack_and_queue/deque.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
+/**
+ * File: deque.js
+ * Created Time: 2023-01-17
+ * Author: Zhuo Qinyue (1403450829@qq.com)
+ */
+
+/* 基于数组方法实现的双端队列 */
+class Deque {
```
Comment (krahets) [inline codes/javascript/chapter_stack_and_queue/deque.js:8]: Recommend removing the class Deque, and directly operating on the Array.

**Review by krahets — COMMENTED** [review]
Thanks! Please address the comments.

File: codes/typescript/chapter_stack_and_queue/deque.ts [hunk 3]
```diff
@@ -1,102 +1,39 @@
 /**
- * File: deque.ts
+ * File: deque.js
```
Comment (krahets) [inline codes/typescript/chapter_stack_and_queue/deque.ts:2]: ts

File: codes/typescript/chapter_stack_and_queue/deque.ts [hunk 4]
```diff
@@ -1,102 +1,39 @@
...
-const size: number = deque.size();
+const size: number = deque.length;
 console.log("双向队列长度 size = " + size);
 
 /* 判断双向队列是否为空 */
-const isEmpty: boolean = deque.isEmpty();
+const isEmpty: boolean = size === 0;
 console.log("双向队列是否为空 = " + isEmpty);
 
 export { };
```
Comment (krahets) [inline codes/typescript/chapter_stack_and_queue/deque.ts:39]: export {};

Comment (zhuoqinyue) [reply codes/typescript/chapter_stack_and_queue/deque.ts]: got it

**Comment by zhuoqinyue** [comment]
please check again @krahets

**Review by krahets — APPROVED** [review]
Thanks for the PR! Now merge it.

I recommended turning on the option "Allow edit by the ..." so I can fine-tune this PR.

**Comment by krahets** [comment]
Moreover, could you help implement `linkedlist_deque.js / ts` as well?

**Comment by krahets** [comment]
Please found the finetuned code at https://github.com/krahets/hello-algo/commit/7d0d3df88602296445d6ccb6b8e49938fe27d2ef

**Comment by zhuoqinyue** [comment]
&nbsp;Of course, my pleasure!



---Original---
From: "Yudong ***@***.***&gt;
Date: Sat, Feb 4, 2023 14:56 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [krahets/hello-algo] Add deque for JavaScript and typescript (PR#267)




 
Moreover, could you help implement linkedlist_deque.js / ts as well?

**Pull request merged**